src/utility/include/mp-units/math.h

```diff
@@ -136,6 +136,30 @@ template<auto R, typename Rep>
   return {static_cast<Rep>(abs(q.numerical_value_ref_in(q.unit))), R};
 }
 
+/**
+ * @brief Computes the fma of 3 quantities
+ *
+ * @param a: Multiplicand
+ * @param x: Multiplicand
+ * @param b: Addend
+ * @return Quantity: The nearest floating point representable to ax+b
+ */
+template<auto R, auto S, auto T, typename Rep>
+[[nodiscard]] constexpr quantity<T, Rep> fma(const quantity<R, Rep>& a, const quantity<S, Rep>& x,
+                                             const quantity<T, Rep>& b) noexcept
+  requires requires {
+    fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit));
+  } || requires {
+    std::fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit));
+  }
+{
+  using std::fma;
+  return {static_cast<Rep>(
+            fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit))),
+          T};
+}
```

Comment:
Please, see <https://cpp1.godbolt.org/z/5xrq836dd>.
- Account for changes to `std::fma` in C++23: <https://en.cppreference.com/w/cpp/numeric/math/fma>.
- Add the requirement on the references: `requires { common_reference(R * S, T); }`.
  This might not be totally correct.
  I suspect they should be the same reference, rather than just having a common one.
  Other math functions convert to the common quantity, but I suspect we don't want that here.
- This suggesting is formatted.
- I can't figure out the right spelling for the `QuantityOf` return type.
```suggestion
template<auto R, auto S, auto T, typename Rep1, typename Rep2, typename Rep3>
[[nodiscard]] constexpr Quantity/*Of<T>*/ auto fma(
    const quantity<R, Rep1>& a, const quantity<S, Rep2>& x, const quantity<T, Rep3>& b) noexcept
  requires requires { common_reference(R * S, T); } &&
  (requires {
    fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit));
  } || requires {
    std::fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit));
  })
{
  using std::fma;
  return quantity{
    fma(a.numerical_value_ref_in(a.unit), x.numerical_value_ref_in(x.unit), b.numerical_value_ref_in(b.unit)), T};
}
```

Reply:
I never used `fma` before. Should `Rep` be constrained to be a floating-point, or should it work for ints as well? If so, `treat_as_floating_point` should be used for the constraint.

Reply:
> I suspect they should be the same reference, rather than just having a common one.

Not really. For example, you can multiply _speed_ and _time_ and add _height_ to the result. A `common_reference` is a correct constraint to use here. Good catch!

Reply:
> I can't figure out the right spelling for the QuantityOf return type.

The return type can be spelled as `quantity<common_reference(R * S, T), Rep>`.

Reply:
> Account for changes to std::fma in C++23

I didn't check C++23 specs. Does it still require the arguments to be of a single homogeneous type, or can we mix floats and doubles in one call?

Reply:
The constraint should be enough, since `std::fma` is itself constrained (or overload resolution fails due to ambiguity).

Please, fix the return type constraint here: <https://cpp1.godbolt.org/z/hP16W3Kab>.

> > I suspect they should be the same reference, rather than just having a common one.
> 
> Not really. For example, you can multiply _speed_ and _time_ and add _height_ to the result. A `common_reference` is a correct constraint to use here. Good catch!

Sorry, I meant to talk about their units.
Should we reject `fma(a * km, b * km, c * m2)`
because ab km² and c m² don't have the same units?
I suspect that passing the raw values to `st::fma` might not be correct
if the units of multiplying the first two arguments is not the same as that of the third argument.

Reply:
Yes, we should reject, for now.
Later, we can choose whether to relax, or convert them to their common unit.

Reply:
> Please, fix the return type constraint here: <https://cpp1.godbolt.org/z/hP16W3Kab>.

@mpusz Also the constraint to require equal units.

Reply:
You could try asserting like this (<https://cpp1.godbolt.org/z/PvnTx7e4h>):
`compare(fma(2.0 * km, 1.0 * km, 1e6 * m2), std::fma(2.0, 1.0, 1e6) * m2)`.

Reply:
No, that conversion isn't happening.
That's what <https://github.com/mpusz/mp-units/pull/526#discussion_r1421740769> is about.

Reply:
> as currently written this statement is not rejected and the test passes. Let me know if you believe it should be rejected.

Yes, I think that @JohelEGP is right here. Adding raw values of kilometers to a meters without calling to a common unit first is wrong, and we can't scale to a common unit as we can't inject such code to be executed between multiplication and addition.

So, we have to constrain the inputs. I would do this in the following way https://cpp1.godbolt.org/z/aY8r95MMT.